### PR TITLE
Added a grouping procedure verification test case.

### PR DESCRIPTION
**Pull request by VitaliiYerys**
Also groups_example.yaml file was added for testing purposses.

### CODE REVIEW

**Comment by Wasapl** [comment]
Build finished.
Unittest success

**Comment by Wasapl** [comment]
Unittest success

**Comment by Wasapl** [comment]
Build finished.
Functional test failure

**Comment by Wasapl** [comment]
Functional test failure

**Comment by Wasapl** [comment]
Build finished.
Functional test failure

**Comment by Wasapl** [comment]
Functional test failure

**Comment by Wasapl** [comment]
Build finished.
Unittest success

**Comment by Wasapl** [comment]
Unittest success

**Comment by Wasapl** [comment]
Build finished.
Functional test success

**Comment by Wasapl** [comment]
Functional test success

**Comment by Wasapl** [comment]
Build finished.
Functional test success

**Comment by Wasapl** [comment]
Functional test success

File: devlab/tests/test_group_verification.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    """
+    Testing class for grouping verification.
+    """
+    def setUp(self):
+        """
+        SetUp method.
+        """
+        self.src_cloud = Prerequisites(cloud_prefix='SRC')
+        self.main_folder = os.path.dirname(os.path.dirname(os.getcwd()))
+        cmd = './../../devlab/provision/generate_config.sh --cloudferry-path ' \
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: Using relative paths is a bad practice. Please rewrite with full path

**Comment by Wasapl** [comment]
Build finished.
Unittest success

**Comment by Wasapl** [comment]
Unittest success

**Comment by Wasapl** [comment]
Build finished.
Functional test failure

**Comment by Wasapl** [comment]
Functional test failure

File: devlab/tests/test_group_verification.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    def _prepare_files(self, grouping_by):
+        """
+        Method for generation of test/configuration files.
+        """
+        self.post_conf_dict = {}
+        self.pre_conf_dict = {}
+        main_folder = self.main_folder
+
+        exmpl_file_path = '{}/devlab/tests/groups_example.yaml'.format(main_folder)
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: use `os.path.join` instead

File: devlab/tests/test_group_verification.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        cmd = './../../devlab/provision/generate_config.sh --cloudferry-path ' \
+              '{}'.format(self.main_folder)
+        os.system(cmd)
+
+    def tearDown(self):
+        """
+        Clean up method.
+        """
+        cmd_1 = 'rm {}/devlab/tests/{}'.format(self.main_folder, self.new_file_name)
+        cmd_2 = 'rm {}/vm_groups.yaml'.format(self.main_folder)
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: use `os.path.join` instead

File: devlab/tests/test_group_verification.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                            inst_id_list.append(inst['id'])
+            elif key == 'user_defined_group_2':
+                for inst in self.src_vms:
+                    if inst['name'] == self.pre_conf_dict[key][0]:
+                        inst_3 = inst['id']
+        self.pre_conf_dict['group_by'] = [unicode(grouping_by)]
+        self.pre_conf_dict['user_defined_group_1'] = inst_id_list
+        self.pre_conf_dict['user_defined_group_2'] = [inst_3]
+        self.new_file_name = 'test_file.yaml'
+        conf_file = open('{}/{}'.format(os.getcwd(), self.new_file_name), 'w')
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: same here

File: devlab/tests/test_group_verification.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        conf_file.write(yaml.dump(self.pre_conf_dict, default_flow_style=False))
+        conf_file.close()
+        cmd = 'cd {} && fab get_groups:configuration.ini,devlab/t' \
+              'ests/{}'.format(main_folder, self.new_file_name)
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: and here

File: devlab/tests/test_group_verification.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        os.system(cmd)
+        post_file_path = '{}/vm_groups.yaml'.format(main_folder)
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: and here

File: devlab/tests/test_group_verification.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    """
+    Testing class for grouping verification.
+    """
+    def setUp(self):
+        """
+        SetUp method.
+        """
+        self.src_cloud = Prerequisites(cloud_prefix='SRC')
+        self.main_folder = os.path.dirname(os.path.dirname(os.getcwd()))
+        # TODO:
```
Comment (realsystem) [inline devlab/tests/test_group_verification.py]: Is there appropriate issue inJira to track this the activity about proper package? If not, please, create.

Comment (VitaliiYerys) [inline devlab/tests/test_group_verification.py]: Task created:

**Comment by VitaliiYerys** [comment]
In order for tests Pass with this patch, Patch #281 has to be merged first.

File: devlab/tests/test_group_verification.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                    if inst['name'] == self.pre_conf_dict[key][0]:
+                        inst_3 = inst['id']
+        self.pre_conf_dict['group_by'] = [unicode(grouping_by)]
+        self.pre_conf_dict['user_defined_group_1'] = inst_id_list
+        self.pre_conf_dict['user_defined_group_2'] = [inst_3]
+        self.new_file_name = 'test_file.yaml'
+        conf_file = open(os.path.join(os.getcwd(), self.new_file_name), 'w')
+        conf_file.write(yaml.dump(self.pre_conf_dict, default_flow_style=False))
+        conf_file.close()
+        fab_path = os.path.join('devlab/tests', self.new_file_name)
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: Please use `with open(...) as f:` instead

File: devlab/tests/test_group_verification.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        cmd = 'cd {} && fab get_groups:configuration.ini,{}'.format(main_folder, fab_path)
+        os.system(cmd)
+        post_file_path = os.path.join(main_folder, 'vm_groups.yaml')
+        post_conf = open(post_file_path, 'r')
+        self.post_conf_dict = yaml.load(post_conf)
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: `yaml.load` requires stream descriptor, would be better to have `yaml.load(file(<filename>, 'r'))`

File: devlab/tests/test_group_verification.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        #  Should be fixed by implementing proper package module for Cloud Ferry.
+        cmd_no_path = './../../devlab/provision/generate_config.sh --cloudferry-path '
+        cmd = os.path.join(cmd_no_path, self.main_folder.lstrip('/'))
+        os.system(cmd)
+
+    def tearDown(self):
+        """
+        Clean up method.
+        """
+        cmd_1 = 'rm {}'.format(os.path.join(self.main_folder, 'devlab/tests', self.new_file_name.lstrip('/')))
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: pep8 violation, line too long

File: devlab/tests/test_group_verification.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                    if inst['name'] == self.pre_conf_dict[key][0]:
+                        inst_3 = inst['id']
+        self.pre_conf_dict['group_by'] = [unicode(grouping_by)]
+        self.pre_conf_dict['user_defined_group_1'] = inst_id_list
+        self.pre_conf_dict['user_defined_group_2'] = [inst_3]
+        self.new_file_name = 'test_file.yaml'
+        #
+        # conf_file = open(os.path.join(os.getcwd(), self.new_file_name), 'w')
+        # conf_file.write(yaml.dump(self.pre_conf_dict, default_flow_style=False))
+        # conf_file.close()
```
Comment (roman-verchikov) [inline devlab/tests/test_group_verification.py]: Is it a debug leftover?.. Should this be removed?

**Comment by Settis** [comment]
+1

**Comment by realsystem** [comment]
+1

**Pull request merged**